### PR TITLE
Drop support for legacy Python 2.7

### LINKED ISSUE #228: Deprecate compability with python 2.7
Hi there,

Recently I run something with Python 2.7 and got the following message: 

"DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support"

I think that by January 1st, we should drop support to Python 2.7 as well. What do you think?

@davehunt @BeyondEvil

### PR DESCRIPTION
**Pull request by hugovk**
Fixes #228.

### CODE REVIEW

File: pytest_html/plugin.py [hunk 1]
```diff
@@ -346,8 +334,7 @@ def _generate_report(self, session):
         self.style_css = pkg_resources.resource_string(
             __name__, os.path.join("resources", "style.css")
         )
-        if PY3:
-            self.style_css = self.style_css.decode("utf-8")
+        self.style_css = self.style_css.decode("utf-8")
```
Comment (BeyondEvil) [inline pytest_html/plugin.py]: You can probably do

```
self.style_css = pkg_resources.resource_string(
    __name__, os.path.join("resources", "style.css")
).decode("utf-8")
```

File: pytest_html/plugin.py [hunk 2]
```diff
@@ -362,12 +349,12 @@ def _generate_report(self, session):
...
             self.style_css += "\n/******************************"
             self.style_css += "\n * CUSTOM CSS"
-            self.style_css += "\n * {}".format(path)
+            self.style_css += f"\n * {path}"
             self.style_css += "\n ******************************/\n\n"
             with open(path, "r") as f:
                 self.style_css += f.read()
 
-        css_href = "{0}/{1}".format("assets", "style.css")
+        css_href = "{}/{}".format("assets", "style.css")
```
Comment (BeyondEvil) [inline pytest_html/plugin.py]:
```suggestion
        css_href = "assets/style.css"
```

File: pytest_html/plugin.py [hunk 3]
```diff
@@ -473,18 +460,17 @@ def generate_summary_item(self):
         main_js = pkg_resources.resource_string(
             __name__, os.path.join("resources", "main.js")
         )
-        if PY3:
-            main_js = main_js.decode("utf-8")
+        main_js = main_js.decode("utf-8")
```
Comment (BeyondEvil) [inline pytest_html/plugin.py]: same as `self.style_css` above.

File: pytest_html/plugin.py [hunk 4]
```diff
@@ -501,11 +487,11 @@ def generate_summary_item(self):
...
-        unicode_doc = u"<!DOCTYPE html>\n{0}".format(doc.unicode(indent=2))
-        if PY3:
-            # Fix encoding issues, e.g. with surrogates
-            unicode_doc = unicode_doc.encode("utf-8", errors="xmlcharrefreplace")
-            unicode_doc = unicode_doc.decode("utf-8")
+        unicode_doc = "<!DOCTYPE html>\n{}".format(doc.unicode(indent=2))
+
+        # Fix encoding issues, e.g. with surrogates
+        unicode_doc = unicode_doc.encode("utf-8", errors="xmlcharrefreplace")
+        unicode_doc = unicode_doc.decode("utf-8")
```
Comment (BeyondEvil) [inline pytest_html/plugin.py]: Probably skip this line and just do `return unicode_doc.decode("utf-8")`

File: tox.ini [hunk 5]
```diff
@@ -4,15 +4,15 @@
 # and then run "tox" from this directory.
 
 [tox]
-envlist = py{27,36,37,py,py3}{,-ansi2html}, linting
+envlist = py{36,37,py,py3}{,-ansi2html}, linting
```
Comment (BeyondEvil) [inline tox.ini]: Probably remove `py` here (as it's 2.7 based)

File: tox.ini [hunk 6]
```diff
@@ -4,15 +4,15 @@
...
 
 [testenv]
 setenv = PYTHONDONTWRITEBYTECODE=1
 deps =
     pytest-xdist
     pytest-rerunfailures
     pytest-mock
-    py{27,36,py,py3}-ansi2html: ansi2html
+    py{36,py,py3}-ansi2html: ansi2html
```
Comment (BeyondEvil) [inline tox.ini]: Probably replace `py` with `37` here.

**Comment by hugovk** [comment]
Thanks for the review, updated!

---

> I'm not sure how to handle the merge however. It is after all a couple of months left before 2.7 is EOL.

To compare with pytest itself: 

* Pytest 5.x is the master branch, and no longer supports Python 2.7.

* The pytest core team will make bug fixes to pytest 4.6 until January 2020, and will accept and release community patches until mid-2020:

* https://docs.pytest.org/en/latest/py27-py34-deprecation.html

File: testing/test_pytest_html.py [hunk 7]
```diff
@@ -30,20 +27,20 @@ def read_html(path):
...
     # Asserts if the test number of this outcome in the summary is correct
-    regex_summary = r"(\d)+ {0}".format(label or test_outcome)
+    regex_summary = r"(\d)+ {}".format(label or test_outcome)
     assert int(re.search(regex_summary, html).group(1)) == test_outcome_number
 
     # Asserts if the generated checkbox of this outcome is correct
     regex_checkbox = (
         '<input checked="true" class="filter" '
-        'data-test-result="{0}"'.format(test_outcome)
+        'data-test-result="{}"'.format(test_outcome)
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]:
```suggestion
        f'data-test-result="{test_outcome}"'
```

File: testing/test_pytest_html.py [hunk 8]
```diff
@@ -30,20 +27,20 @@ def read_html(path):
...
     )
     if test_outcome_number == 0:
         regex_checkbox += ' disabled="true"'
     assert re.search(regex_checkbox, html) is not None
 
     # Asserts if the table rows of this outcome are correct
-    regex_table = 'tbody class="{0} '.format(test_outcome)
+    regex_table = f'tbody class="{test_outcome} '
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]: Not your fault, but isn't this missing a `"` ?
```suggestion
    regex_table = f'tbody class="{test_outcome}"'
```

File: testing/test_pytest_html.py [hunk 9]
```diff
@@ -280,8 +275,8 @@ def test_stdout(self, testdir, result):
         testdir.makepyfile(
             """
             def test_stdout():
-                print('{0}')
-                assert '{1}' == 'pass'""".format(
+                print('{}')
+                assert '{}' == 'pass'""".format(
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]: Maybe change these to the `f`-style formatting?

File: testing/test_pytest_html.py [hunk 10]
```diff
@@ -299,9 +294,9 @@ def test_custom_content_in_summary(self, testdir):
             from py.xml import html
 
             def pytest_html_results_summary(prefix, summary, postfix):
-                prefix.append(html.p("prefix is {0}"))
-                summary.extend([html.p("extra summary is {1}")])
-                postfix.extend([html.p("postfix is {2}")])
+                prefix.append(html.p("prefix is {}"))
+                summary.extend([html.p("extra summary is {}")])
+                postfix.extend([html.p("postfix is {}")])
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]: Same here.

File: testing/test_pytest_html.py [hunk 11]
```diff
@@ -324,7 +319,7 @@ def pytest_runtest_makereport(item, call):
                 report = outcome.get_result()
                 if report.when == 'call':
                     from pytest_html import extras
-                    report.extra = [extras.html('<div>{0}</div>')]
+                    report.extra = [extras.html('<div>{}</div>')]
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]: And here.

File: testing/test_pytest_html.py [hunk 12]
```diff
@@ -348,16 +343,16 @@ def pytest_runtest_makereport(item, call):
                 report = outcome.get_result()
                 if report.when == 'call':
                     from pytest_html import extras
-                    report.extra = [extras.text({0})]
+                    report.extra = [extras.text({})]
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]: dito

File: testing/test_pytest_html.py [hunk 13]
```diff
@@ -371,7 +366,7 @@ def pytest_runtest_makereport(item, call):
                 report = outcome.get_result()
                 if report.when == 'call':
                     from pytest_html import extras
-                    report.extra = [extras.json({0})]
+                    report.extra = [extras.json({})]
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]: yep

File: testing/test_pytest_html.py [hunk 14]
```diff
@@ -399,15 +391,15 @@ def pytest_runtest_makereport(item, call):
                 report = outcome.get_result()
                 if report.when == 'call':
                     from pytest_html import extras
-                    report.extra = [extras.url('{0}')]
+                    report.extra = [extras.url('{}')]
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]: You get the idea. :)

File: testing/test_pytest_html.py [hunk 15]
```diff
@@ -430,16 +422,16 @@ def pytest_runtest_makereport(item, call):
                 report = outcome.get_result()
                 if report.when == 'call':
                     from pytest_html import extras
-                    report.extra = [extras.{0}('{1}')]
+                    report.extra = [extras.{}('{}')]
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]: Oh look, yet another place! 😛

File: testing/test_pytest_html.py [hunk 16]
```diff
@@ -460,7 +452,7 @@ def pytest_runtest_makereport(item, call):
                 report = outcome.get_result()
                 if report.when == 'call':
                     from pytest_html import extras
-                    report.extra = [extras.text({0})]
+                    report.extra = [extras.text({})]
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]: 🙄

File: testing/test_pytest_html.py [hunk 17]
```diff
@@ -488,7 +480,7 @@ def pytest_runtest_makereport(item, call):
                 report = outcome.get_result()
                 if report.when == 'call':
                     from pytest_html import extras
-                    report.extra = [extras.{0}('{1}')]
+                    report.extra = [extras.{}('{}')]
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]: ping

File: testing/test_pytest_html.py [hunk 18]
```diff
@@ -518,7 +510,7 @@ def pytest_runtest_makereport(item, call):
...
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]: pong

File: testing/test_pytest_html.py [hunk 19]
```diff
@@ -552,7 +544,7 @@ def pytest_runtest_makereport(item, call):
                 report = outcome.get_result()
                 if report.when == 'call':
                     from pytest_html import extras
-                    report.extra = [extras.image('{0}')]
+                    report.extra = [extras.image('{}')]
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]: lorum

File: testing/test_pytest_html.py [hunk 20]
```diff
@@ -581,16 +573,16 @@ def pytest_runtest_makereport(item, call):
         test_name = "test_{}".format("a" * 300)
         testdir.makepyfile(
             """
-            def {0}():
+            def {}():
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]: ipsum

File: testing/test_pytest_html.py [hunk 21]
```diff
@@ -635,7 +627,7 @@ def test_environment(self, testdir):
         testdir.makeconftest(
             """
             def pytest_configure(config):
-                config._metadata['content'] = '{0}'
+                config._metadata['content'] = '{}'
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]: I'm running out of memes

File: testing/test_pytest_html.py [hunk 22]
```diff
@@ -652,7 +644,7 @@ def test_environment_xdist(self, testdir):
             """
             def pytest_configure(config):
                 for i in range(2):
-                    config._metadata['content'] = '{0}'
+                    config._metadata['content'] = '{}'
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]: I am your father?

File: testing/test_pytest_html.py [hunk 23]
```diff
@@ -669,7 +661,7 @@ def test_environment_xdist_reruns(self, testdir):
...
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]: Maybe you prefer "live long and prosper"?

File: testing/test_pytest_html.py [hunk 24]
```diff
@@ -683,13 +675,13 @@ def pytest_configure(config):
...
-        expected_content = ", ".join((str(i) for i in content))
-        expected_html_re = r"<td>content</td>\n\s+<td>{}</td>".format(expected_content)
+        expected_content = ", ".join(str(i) for i in content)
+        expected_html_re = fr"<td>content</td>\n\s+<td>{expected_content}</td>"
         testdir.makeconftest(
             """
             def pytest_configure(config):
                 for i in range(2):
-                    config._metadata['content'] = {0}
+                    config._metadata['content'] = {}
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]: how meta

File: testing/test_pytest_html.py [hunk 25]
```diff
@@ -786,7 +772,7 @@ def pytest_runtest_makereport(item, call):
                 outcome = yield
                 report = outcome.get_result()
                 if report.when == 'call':
-                    report.longrepr = 'utf8 longrepr: ' + {0}
+                    report.longrepr = 'utf8 longrepr: ' + {}
```
Comment (BeyondEvil) [inline testing/test_pytest_html.py]: last one, yay!

**Comment by BeyondEvil** [comment]
Good job on the changes! Discovered some more, sorry about that.

> Thanks for the review, updated!
> 
> > I'm not sure how to handle the merge however. It is after all a couple of months left before 2.7 is EOL.
> 
> To compare with pytest itself:
> 
>     * Pytest 5.x is the master branch, and no longer supports Python 2.7.
> 
>     * The pytest core team will make bug fixes to pytest 4.6 until January 2020, and will accept and release community patches until mid-2020:
> 
>     * https://docs.pytest.org/en/latest/py27-py34-deprecation.html

Not a bad idea. I guess we can have a `twodotseven`-branch for the same purpose.

But I want someone else to chime in before I make it so. @davehunt @rochacbruno

File: pytest_html/plugin.py [hunk 26]
```diff
@@ -276,7 +262,7 @@ def append_log_html(self, report, additional_html):
 
             for section in report.sections:
                 header, content = map(escape, section)
-                log.append(" {0} ".format(header).center(80, "-"))
+                log.append(f" {header} ".center(80, "-"))
```
Comment (rochacbruno) [inline pytest_html/plugin.py]: We can replace the `center` with `f-string` formatting.
```suggestion
                log.append(f" {header:-^80} ")
```

as in

```py
>>> f"{'hello':-^15}"                                                                                                                                                                         
'-----hello-----'
```

File: pytest_html/plugin.py [hunk 27]
```diff
@@ -296,7 +282,7 @@ def _appendrow(self, outcome, report):
             self.results.insert(index, result)
             tbody = html.tbody(
                 result.row_table,
-                class_="{0} results-table-row".format(result.outcome.lower()),
+                class_="{} results-table-row".format(result.outcome.lower()),
```
Comment (rochacbruno) [inline pytest_html/plugin.py]: we can call functions from inside the `f-string` what do you think about it?

```suggestion
                class_=f"{result.outcome.lower()} results-table-row",
```

File: pytest_html/plugin.py (hunk 26 above)
Comment (BeyondEvil) [reply pytest_html/plugin.py]: Nice!

File: pytest_html/plugin.py (hunk 27 above)
Comment (BeyondEvil) [reply pytest_html/plugin.py]: My personal preference is to use `f-string` for "pure" variables, and `format` for everything else (dicts, calling functions etc).

To me, the function call becomes a little bit obfuscated when "hidden away" in the f-string.

But if you feel strongly about it, I'm fine with having it that way. 😊

**Comment by RibeiroAna** [comment]
@BeyondEvil my opinion is that we should launch Pytest-HTML 2.0 with this patch and people who still need to use Python 2.7 during the remaining months until its dead should use Pytest-HTML 1.x

**Comment by BeyondEvil** [comment]
> @BeyondEvil my opinion is that we should launch Pytest-HTML 2.0 with this patch and people who still need to use Python 2.7 during the remaining months until its dead should use Pytest-HTML 1.x

I agree. I'm just not sure how to handle the practical aspects.

**Comment by RibeiroAna** [comment]
@BeyondEvil I'm not sure which practical aspects you mean! Maybe launch Pytest 1.23.0 that includes a deprecation alert on the HTML file? (I don't think it is worthy)

**Comment by BeyondEvil** [comment]
> @BeyondEvil I'm not sure which practical aspects you mean! Maybe launch Pytest 1.23.0 that includes a deprecation alert on the HTML file? (I don't think it is worthy)

Well, if master is going to be 2.x, we'll need a 1.x-"master"-branch, and I'm not sure if either of us can create that or if only @davehunt can do it.

And I agree, not worth releasing a 1.x version with just a deprecation warning. `pytest` itself already does that.

File: setup.py [hunk 28]
```diff
@@ -15,6 +15,7 @@
     install_requires=["pytest>=3.0", "pytest-metadata"],
```
Comment (BeyondEvil) [inline setup.py]:
```suggestion
    install_requires=["pytest>=5.0", "pytest-metadata"],
```

**Comment by hugovk** [comment]
Thanks for the new reviews! I'm travelling right now and can update these later this week. Or feel free to update this branch. 

The `python_requires=">=3.6",` here helps with some of the practical aspects. It means people (with modern pip) who install the package will only get this version if their Python version matches. Otherwise pip'll check each older one until it finds one that matches (or has no python_requires). So 2.7 users will get the last one to support 2.7.

https://python3statement.org/practicalities/ has some other suggestions, but I think python_requires is the most important.

**Comment by BeyondEvil** [comment]
@hugovk No worries, there's no stress! We're not fixing anything that is broken.

Safe travels!

File: testing/test_pytest_html.py (hunk 8 above)
Comment (hugovk) [reply testing/test_pytest_html.py]: Could be. Changing it fails tests, so should be dealt with separately.

**Comment by hugovk** [comment]
Updated!

**Comment by davehunt** [comment]
I'd say let's release this as pytest-html 2.0 and create a 1.22-maintenance branch, following the same approach as pytest, and only uplifting bugfixes to that branch and releasing 1.22.x as needed. I don't expect there will be many (if any) patches that will require a 1.22.x release though, and we should strongly encourage users to upgrade to Python 3.

**Pull request merged**